File: code/109-Convert Sorted List to Binary Search Tree/109-Convert Sorted List to Binary Search Tree.py
# -*- coding: utf-8 -*
"""
@author: Kai Chen
@file: 109-Convert Sorted List to Binary Search Tree.py
@time: 2020/8/18 9:57
@desc:
Given the head of a singly linked list where elements are sorted in ascending order, convert it to a height balanced BST.

For this problem, a height-balanced binary tree is defined as a binary tree in which the depth of the two subtrees of every node never differ by more than 1.



Example 1:


Input: head = [-10,-3,0,5,9]
Output: [0,-3,9,-10,null,5]
Explanation: One possible answer is [0,-3,9,-10,null,5], which represents the shown height balanced BST.
Example 2:

Input: head = []
Output: []
Example 3:

Input: head = [0]
Output: [0]
Example 4:

Input: head = [1,3]
Output: [3,1]


Constraints:

The numner of nodes in head is in the range [0, 2 * 10^4].
-10^5 <= Node.val <= 10^5
"""


# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution(object):
    def sortedListToBST(self, head):
        """
        :type head: ListNode
        :rtype: TreeNode
        """

        def getMedian(left, right):
            fast = slow = left
            while fast != right and fast.next != right:
                fast = fast.next.next
                slow = slow.next
            return slow
        def buildTree(left, right):
            if left == right:
                return None
            mid = getMedian(left, right)
            root = TreeNode(mid.val)
            root.left = buildTree(left, mid)
            root.right = buildTree(mid.next, right)
            return root
        return buildTree(head, None)

File: code/109-Convert Sorted List to Binary Search Tree/109-Convert Sorted List to Binary Search Tree.cpp
/*
LeetCode 109 - Convert Sorted List to Binary Search Tree

解题步骤（自动整理）
1. 根据题意选择前序/中序/后序遍历方式遍历二叉树
2. 在遍历过程中维护需要的状态（路径和、深度、父子关系等）
3. 遍历完成后返回答案
*/
/*Convert Sorted List to Binary Search Tree
Given a singly linked list where elements are sorted in ascending order, convert it to a height balanced BST.
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 * //List不能随机访问，所以这个办法是写一个按照index访问list的办法，以及计算List长度的办法，时间复杂度O(n^2)
 */
class Solution {
public:
    TreeNode *sortedListToBST(ListNode *head) {
        return sortedListToBST (head, listLength (head));
    }

    TreeNode *sortedListToBST (ListNode *head, int length){
        if(length==0) return NULL;
        if(length==1) return new TreeNode(head->val);

        TreeNode *root = new TreeNode(nth_node(head,length/2+1)->val);
        root->left = sortedListToBST(head, length/2);
        root->right= sortedListToBST(nth_node(head, length/2+2), (length-1)/2);

        return root;
    }

    int listLength(ListNode *node){

        int n=0;
        while(node){
            n++;
            node=node->next;
        }
        return n;
    }

    ListNode *nth_node(ListNode *node, int n){
        while(--n){//n==1，时候就可以return node了
          node=node->next;
        }

        return node;
    }
};
